src/connection_pool.rs

```diff
@@ -107,36 +113,175 @@ where
             last_error: None,
             resolver_thread: Some(resolver_thread),
             resolver_rx_thread: Some(resolver_rx_thread),
-            spares: cpo.spares,
+            resolver_tx: Some(tx),
             maximum: cpo.maximum,
             claim_timeout: cpo.claim_timeout,
             rebalance_check: rebalancer_check,
             rebalance_thread: Some(rebalance_thread),
+            rebalancer_stop,
             log: cpo.log,
+            state: ConnectionPoolState::Running,
             _resolver: PhantomData
         };
 
         barrier.clone().wait();
         pool
     }
 
-    pub fn stop(&self) {
-        std::unimplemented!()
+
+    /// Stop the connection pool and resolver and close all connections in a
+    /// graceful manner. This function may only be called on the original
+    /// ConnectionPool instance. Thread JoinHandles may not be cloned and
+    /// therefore invocation of this function by a clone of the pool results in
+    /// an error. This function will block the caller until all claimed threads
+    /// are returned to the pool and are closed and until all worker threads
+    /// except for the thread running the resolver have exited.
+    pub fn stop(&mut self) -> Result<(), Error> {
+        // Make sure this is the original ConnectionPool instance and that it is
+        // safe to unwrap all the Option types needed to stop the pool.
+        if self.resolver_thread.is_some() &&
+            self.resolver_rx_thread.is_some() &&
+            self.resolver_tx.is_some() &&
+            self.rebalance_thread.is_some()
+        {
+            // Transition state to Stopping
+            self.state = ConnectionPoolState::Stopping;
+
+            // Notify the resolver, resolver_recv, and rebalancer threads to
+            // shutdown.  Join on the thread handles for the resolver receiver
+            // and rebalancer threads. Do not join on the resolver thread
+            // because the code is not controlled by the pool and may not
+            // properly respond to the resolver receiver thread stopping. Just
+            // drop the JoinHandle for the resolver thread and move along.
+            self.rebalancer_stop.store(true, AtomicOrdering::Relaxed);
+            let resolver_tx = self.resolver_tx.take().unwrap();
+            match resolver_tx.send(BackendMsg::StopMsg) {
+                Ok(()) => {
+                    let resolver_rx_thread = self.resolver_rx_thread.take().unwrap();
+                    let _ = resolver_rx_thread.join();
+                },
+                Err(e) => {
+                    warn!(self.log, "Failed to send stop message to resolver \
+                                     receiver thread: {}", e);
+                }
+            }
+            drop(resolver_tx);
+
+            let rebalance_thread = self.rebalance_thread.take().unwrap();
+            let _ = rebalance_thread.join();
+            drop(self.resolver_thread.take().unwrap());
```

Comment:
Is it possible for this to leave the resolver running after that connection pool stops? I think we discussed a signaling mechanism to shut down the resolver, or will this eventually call the stop() via the Resolver trait?

Reply:
The pool does not maintain a channel to send direct messages to the resolver so the signal for a resolver to stop is when the `rx` side of the resolver channel is dropped. That happens when the `resolver_rx_thread` stops and the `rx` half of the channel goes out of scope. After that any attempt by the resolver to send a message back to the pool will result in an error and this should be used by resolvers as the signal to shut down. So it is the case that a resolver could continue running if it ignores this signal, but it would effectively be cut off from the world.

Reply:
Great explanation, thanks. Will keep that in mind on the resolver side.